aten/src/ATen/Device.h

```diff
@@ -112,3 +113,16 @@ struct Device {
 
 std::ostream& operator<<(std::ostream& stream, at::Device::Type type);
 std::ostream& operator<<(std::ostream& stream, const at::Device& device);
+
+namespace std {
+  template<> struct hash<at::Device>
+  {
+    int64_t operator()(const at::Device& device) const noexcept {
+      int64_t hash_val = static_cast<int64_t>(device.index());
+      if (device.type() == at::Device::Type::CUDA) {
+        hash_val += 2;
```

Comment:
The logic is simple: permissible values for CPU indices are -1, 0. CUDA indices can be arbitrary, starting from -1. CUDA hash values will hence be index + 2.

However, this still causes non-determinism when CUDA device index > 254. [This](https://docs.python.org/3/c-api/long.html#c.PyLong_FromLong) could be the reason.

Reply:
Seems very likely. If that is true, we shouldn't depend on this UB really.

Reply:
The upper bound is still high practically speaking.

Another simple way to deterministically hash it would be to use the device string:
```
std::string device_string = toString(device.type()) + ":" + to_string(device.index());
return std::hash<std::string>{}(device_string);
```
But again, this value might result in non-deterministic overflow.